src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -282,6 +282,16 @@ private static string GenerateUpgradeReport(string projectName, string backupPat
             }
         }
 
+        private static void AddToExisting(IPackageReferenceContextInfo pkg, ISet<Tuple<string, string>> existingPackages)
+        {
+            PackageIdentity package = pkg.Identity;
+            Tuple<string, string> packageInfo = Tuple.Create(package.Id, package.Version == null ? "" : package.Version.ToNormalizedString());
+            if (!existingPackages.Contains(packageInfo))
+            {
+                existingPackages.Add(packageInfo);
+            }
+        }
```

Comment:
This method could be replaced with a method that creates a tuple from the package identity and returns it. That way, the method does one thing and does it well. Then, the calling code can just add that tuple to the existingPackage hash set. Note that we do not need to ask a hash set if it contains an item before adding it to the hash set. We can just call set.Add() and try to add a new item and if it already exists, it just won't add it again.

```suggestion
        private static Tuple<string, string> CreatePackageTuple(IPackageReferenceContextInfo pkg)
        {
            PackageIdentity package = pkg.Identity;
            return Tuple.Create(package.Id, package.Version == null ? string.Empty : package.Version.ToNormalizedString());
        }
```

Then, the caller of this method would simply call:
`existingPackages.Add(CreatePackageTuple(package));`

Reply:
Added to code. Thanks!